drivers/common/virtio/virtio.c

```diff
@@ -0,0 +1,339 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright (c) 2022 NVIDIA Corporation & Affiliates
+ */
+
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <rte_malloc.h>
+#include <rte_vfio.h>
+
+#include "virtio_pci.h"
+#include "virtio_api.h"
+#include "virtqueue.h"
+#include "virtio_logs.h"
+
+struct virtio_pci_dev *
+virtio_dev_alloc(struct rte_pci_device *pci_dev)
+{
+	struct virtio_pci_dev *vpdev;
+	struct virtio_hw *hw = NULL;
+	int ret;
+
+	vpdev = rte_zmalloc("virtio pci device", sizeof(*vpdev), RTE_CACHE_LINE_SIZE);
+	if (vpdev == NULL) {
+		PMD_INIT_LOG(ERR, "Failed to allocate vpdev memory.");
+		return NULL;
+	}
+
+	hw = &vpdev->hw;
+
+	VTPCI_DEV(hw) = pci_dev;
+	ret = virtio_dev_pci_init(pci_dev, vpdev);
+	if (ret) {
+		PMD_INIT_LOG(ERR, "Failed to init virtio PCI device");
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	vpdev->vfio_dev_fd = rte_intr_dev_fd_get(pci_dev->intr_handle);
+	if (vpdev->vfio_dev_fd < 0) {
+		PMD_INIT_LOG(ERR, "Failed to get vfio dev fd");
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	/* Reset the device although not necessary at startup */
+	virtio_dev_reset(vpdev);
+
+	/* Tell the host we've noticed this device. */
+	virtio_dev_set_status(vpdev, VIRTIO_CONFIG_STATUS_ACK);
+
+	/* Tell the host we've known how to drive the device. */
+	virtio_dev_set_status(vpdev, VIRTIO_CONFIG_STATUS_DRIVER);
+
+	hw->guest_features = VIRTIO_OPS(hw)->get_features(hw);
+	PMD_INIT_LOG(DEBUG,"Guest_features is 0x%"PRIx64,hw->guest_features);
+	return vpdev;
+
+error:
+		if (vpdev) {
+			rte_free(vpdev);
+		}
+		return NULL;
+
+}
+
+void
+virtio_dev_free(struct virtio_pci_dev *vpdev)
+{
+	struct virtio_hw *hw = NULL;
+
+	hw = &vpdev->hw;
+	if (VIRTIO_OPS(hw)->dev_close(hw))
+		PMD_INIT_LOG(ERR, "Failed to close virtio device.");
+	rte_free(vpdev);
+}
+
+uint16_t
+virtio_dev_nr_vq_get(struct virtio_pci_dev *vpdev)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+	return VIRTIO_OPS(hw)->get_queue_num(hw);
+}
+
+void
+virtio_dev_free_queues(struct virtio_pci_dev *vpdev,uint16_t nr_vq)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+	struct virtqueue *vq;
+	uint16_t i;
+
+	if (hw->vqs == NULL)
+		return;
+
+	for (i = 0; i < nr_vq; i++) {
+		vq = hw->vqs[i];
+		if (!vq)
+			continue;
+		rte_free(vq);
+		hw->vqs[i] = NULL;
+	}
+
+	rte_free(hw->vqs);
+	hw->vqs = NULL;
+}
+
+int
+virtio_dev_alloc_queues(struct virtio_pci_dev *vpdev,uint16_t nr_vq)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+	struct virtqueue *vq;
+	uint16_t i;
+
+	hw->vqs = rte_zmalloc(NULL, sizeof(struct virtqueue *) * nr_vq, 0);
```

Comment:
do sizeof(vq) instead of sizeof(sizeof(struct virtqueue *)).

This ensures that any structure name change etc doesn't affect the code tomorrow.
it will be always size of vq structure regardless of its type.